benchmarks/upload_metrics_to_bq.py

```diff
@@ -0,0 +1,230 @@
+  last_n_data = []
+  with open(metrics_file, 'r', encoding='utf8') as file:
+    lines = file.readlines()
+    for line in lines[::-1]:
+      metrics = json.loads(line)
+      if target in metrics:
+        last_n_data.append(metrics[target])
+        if len(last_n_data) >= n:
+          break
+  return last_n_data
+
+# Get the average of the last n datapoints for a specific target metric
+def get_metric_average(metrics_file, target, n=10):
+  last_n_data = get_last_n_data(metrics_file, target, n=n)
+  return sum(last_n_data) / len(last_n_data)
+
+
+def get_metric_median(metrics_file, target, n=10):
+  last_n_data = get_last_n_data(metrics_file, target, n=n)
+  return median(last_n_data)
+
+def get_metrics_sum(metrics_file, target, n=10):
+  last_n_data = get_last_n_data(metrics_file, target, n=n)
+  return sum(last_n_data)
+
+# metric file example:
+# {"learning/grad_norm": 1.0000004768371582, "learning/loss": 10.8693265914917, "learning/moe_lb_loss": 0.0, "learning/param_norm": 8122.166015625, "learning/raw_grad_norm": 5.087409973144531, "learning/total_weights": 79694.0, "perf/step_time_seconds": 4.986376, "perf/per_device_tflops": 172.941153140736, "perf/per_device_tflops_per_sec": 34.6827341421377, "perf/per_device_tokens": 24576.0, "perf/per_device_tokens_per_sec": 4928.629529742643, "learning/current_learning_rate": 2.9999999242136255e-05, "step": 0.0, "run_name": "mattdavidow-train-base"}
+
+# It is important to parse the last n steps avoiding steps that have tracing turned on, by default last_n_steps should be total_steps - 10 to avoid initial tracing settings.
+# TODO() Support avoiding tracing when the user explicitly enables tracing for other steps.
+def parse_metrics(local_metrics_file, total_steps, last_n_steps=10) -> Metrics:
+  avg_tflops = get_metric_average(local_metrics_file, "perf/per_device_tflops_per_sec", n=last_n_steps)
+  avg_tokens_per_second = get_metric_average(local_metrics_file, "perf/per_device_tokens_per_sec", n=last_n_steps)
+  median_step_time = get_metric_median(local_metrics_file, "perf/step_time_seconds", n=last_n_steps)
+  e2e_step_time = get_metrics_sum(local_metrics_file, "perf/step_time_seconds", n=total_steps)
+
+  metrics = Metrics(
+    avg_tflops_per_sec=avg_tflops,
+    avg_tokens_per_sec=avg_tokens_per_second,
+    median_step_time=median_step_time,
+    e2e_step_time=e2e_step_time)
+
+  return metrics
+
+
+def update_config_with_tuning_params(base_config: omegaconf.DictConfig,
+                                     tuning_params: Dict[str, Any]):
+  """Updates base_config with key-value pairs from tuning_params."""
+  if tuning_params:
+    for key, value in tuning_params.items():
+      omegaconf.OmegaConf.update(base_config, key, value, merge=True)
+  return base_config
+
+
+def main(argv: Sequence[str]) -> None:
+  # only write once
+  if int(os.environ["TPU_WORKER_ID"]) != 0:
+    return
+
+  for arg in argv:
+    print(f"got arg: {arg}")
+
+  metrics_gcs_file = argv[0]
```

Comment:
this feels a bit fragile? could use argparse or something? or even tuple unpacking might be a tiny bit safer